python_src/program_markers/instrumenter.py

```diff
@@ -196,30 +197,114 @@ def get_constant_macro(self) -> str:
+    @staticmethod
+    @abstractmethod
+    def make_preprocessor_code(marker: DCEMarker) -> str:
+        """Generates the preprocessor code of this strategy for the marker
+
+        Args:
+            marker (DCEMarker):
+                the marker for which the preprocessor code should be generated
+
+        Returns:
+            str:
+                the marker preprocessor code
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def marker_detection_regex() -> re.Pattern[str]:
+        """Returns a pattern of the regex used to detect markers of this strategy
+
+        Returns:
+            re.Pattern[str]:
+                the regex pattern
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def regex_marker_id_group_index() -> int:
+        """Returns the index of the group in the regex, generated by
+        marker_detection_regex(), containing the marker id
+
+        Returns:
+            int:
+                the index of the id group
+        """
+        raise NotImplementedError
+
+
+class MethodCallStrategy(MarkerStrategy):
+    @staticmethod
+    def name() -> str:
+        return "Method call"
+
+    @staticmethod
+    def make_preprocessor_code(marker: DCEMarker) -> str:
+        return f"{marker.marker}();\n void {marker.marker}(void);"
+
+    @staticmethod
+    def marker_detection_regex() -> re.Pattern[str]:
+        return re.compile(f".*(call|jmp).*{DCEMarker.prefix}([0-9]+)_.*")
+
+    @staticmethod
+    def regex_marker_id_group_index() -> int:
+        return 2
+
+
+def find_non_eliminated_markers_impl(
+    asm: str, marker_strategies: tuple[MarkerStrategy, ...]
+) -> tuple[Marker, ...]:
+    """Finds non-eliminated markers of the types specified by `marker_strategies` in
+    `asm`.
+
+    The DCE markers are found by matching the regex pattern specified by the given
+    strategies.
+    The VR markers are found by searching for calls or jumps to functions with
+    names starting with a known marker prefix (e.g., VRMarkerE123_)
 
     Returns:
         tuple[Marker, ...]:
             The non-eliminated markers for the given compilation setting.
     """
-    dce_non_eliminated_regex = re.compile(
-        f".*[call|jmp].*{DCEMarker.prefix}([0-9]+)_.*"
-    )
     vr_non_eliminated_regex = re.compile(
         f".*[call|jmp].*{VRMarker.prefix}([G|L])E([0-9]+)_.*"
     )
     non_eliminated_markers: set[Marker] = set()
     for line in asm.split("\n"):
-        if m := dce_non_eliminated_regex.match(line.strip()):
-            non_eliminated_markers.add(DCEMarker(f"{DCEMarker.prefix}{m.group(1)}_"))
-        elif m := vr_non_eliminated_regex.match(line.strip()):
+        if m := vr_non_eliminated_regex.match(line.strip()):
             non_eliminated_markers.add(
                 VRMarker.from_str(f"{VRMarker.prefix}{m.group(1)}E{m.group(2)}_")
             )
+            continue
+        for marker_strategy in marker_strategies:
```

Comment:
Can we actually end up in a scenario where different markers use different strategies?